subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java

```diff
@@ -0,0 +1,138 @@
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.artifacts.transform.TransformParameters;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.cache.GlobalCacheLocations;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.classpath.InPlaceClasspathBuilder;
+import org.gradle.internal.classpath.transforms.ClasspathElementTransform;
+import org.gradle.internal.classpath.transforms.ClasspathElementTransformFactory;
+import org.gradle.internal.classpath.transforms.ClasspathElementTransformFactoryForAgent;
+import org.gradle.internal.classpath.transforms.ClasspathElementTransformFactoryForLegacy;
+import org.gradle.internal.classpath.transforms.InstrumentingClassTransform;
+import org.gradle.internal.classpath.types.InstrumentingTypeRegistry;
+import org.gradle.internal.file.Stat;
+import org.gradle.util.internal.GFileUtils;
+import org.gradle.work.DisableCachingByDefault;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+import static org.gradle.api.internal.initialization.transform.BaseInstrumentingArtifactTransform.InstrumentArtifactTransformParameters;
+import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTED_JAR_DIR_NAME;
+import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTED_MARKER_FILE_NAME;
+import static org.gradle.internal.classpath.TransformedClassPath.ORIGINAL_JAR_DIR_NAME;
+
+/**
+ * Base artifact transform that instruments plugins with Gradle instrumentation, e.g. for configuration cache detection or property upgrades.
+ */
+@DisableCachingByDefault(because = "Instrumented jars are too big to cache")
+public abstract class BaseInstrumentingArtifactTransform implements TransformAction<InstrumentArtifactTransformParameters> {
+
+    public interface InstrumentArtifactTransformParameters extends TransformParameters {
+        @Input
+        Property<Boolean> getAgentSupported();
+
+        @Input
+        Property<Integer> getMaxSupportedJavaVersion();
+
+        @Input
+        @Optional
+        Property<String> getUpgradedPropertiesHash();
+    }
+
+    @Inject
+    public abstract ObjectFactory getObjects();
+
+    @PathSensitive(PathSensitivity.NAME_ONLY)
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInput();
+
+    @Override
+    public void transform(TransformOutputs outputs) {
+        File input = getInput().get().getAsFile();
+        if (!input.exists()) {
+            // Don't instrument files that don't exist, these could be files added to classpath via files()
+            return;
+        }
```

Comment:
🤔 Hah! Did we inherit this? Feels like something we could filter earlier, or disallow.

Reply:
Is there something we can capture of that learning in a comment here for the next person that tries to remove this?

Reply:
Improved the comment